### PR TITLE
Fix make quick-release-images to work on arm64 and ppc64le

### PR DESCRIPTION
**Pull request by dims**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

- kube-cross should build arch specific images. Starting with arm64 and
ppc64le as etcd binaries are available for these platforms in addition
to usual amd64
- add a manifest to wrap up the arch specific images

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by dims** [comment]
```
openlab@node-0:~/go/src/k8s.io/kubernetes$ make quick-release-images
+++ [0306 16:40:30] Verifying Prerequisites....
+++ [0306 16:40:32] Building Docker image kube-build:build-515f788bad-5-v1.12.0-1
+++ [0306 16:41:23] Creating data container kube-build-data-515f788bad-5-v1.12.0-1
+++ [0306 16:41:43] Syncing sources to container
+++ [0306 16:41:49] Running build command...
+++ [0306 16:42:05] Building go targets for linux/arm64:
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
+++ [0306 16:42:24] Building go targets for linux/arm64:
    ./vendor/k8s.io/code-generator/cmd/defaulter-gen
+++ [0306 16:42:36] Building go targets for linux/arm64:
    ./vendor/k8s.io/code-generator/cmd/conversion-gen
+++ [0306 16:43:01] Building go targets for linux/arm64:
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
+++ [0306 16:43:17] Building go targets for linux/arm64:
    ./vendor/github.com/jteeuwen/go-bindata/go-bindata
+++ [0306 16:43:19] Building go targets for linux/arm64:
    cmd/cloud-controller-manager
    cmd/kube-apiserver
    cmd/kube-controller-manager
    cmd/kube-scheduler
    cmd/kube-proxy
    cmd/hyperkube
    vendor/github.com/onsi/ginkgo/ginkgo
    test/e2e/e2e.test
    cmd/kubectl
+++ [0306 16:49:16] Syncing out of container
+++ [0306 16:49:34] Building images: linux-arm64
+++ [0306 16:49:35] Starting docker build for image: cloud-controller-manager-arm64
+++ [0306 16:49:35] Starting docker build for image: kube-apiserver-arm64
+++ [0306 16:49:35] Starting docker build for image: kube-controller-manager-arm64
+++ [0306 16:49:35] Starting docker build for image: kube-scheduler-arm64
+++ [0306 16:49:35] Starting docker build for image: kube-proxy-arm64
+++ [0306 16:49:35] Building hyperkube image for arch: arm64
+++ [0306 16:49:35] Building conformance image for arch: arm64
+++ [0306 16:49:53] Deleting docker image k8s.gcr.io/kube-scheduler-arm64:v1.15.0-alpha.0.916_4287e699fada74
+++ [0306 16:49:53] Deleting docker image k8s.gcr.io/kube-proxy-arm64:v1.15.0-alpha.0.916_4287e699fada74
+++ [0306 16:49:57] Deleting docker image k8s.gcr.io/cloud-controller-manager-arm64:v1.15.0-alpha.0.916_4287e699fada74
+++ [0306 16:49:58] Deleting docker image k8s.gcr.io/kube-controller-manager-arm64:v1.15.0-alpha.0.916_4287e699fada74
+++ [0306 16:50:01] Deleting docker image k8s.gcr.io/kube-apiserver-arm64:v1.15.0-alpha.0.916_4287e699fada74
+++ [0306 16:50:43] Deleting hyperkube image k8s.gcr.io/hyperkube-arm64:v1.15.0-alpha.0.916_4287e699fada74
+++ [0306 16:50:55] Deleting conformance image k8s.gcr.io/conformance-arm64:v1.15.0-alpha.0.916_4287e699fada74
+++ [0306 16:50:56] Docker builds done
```

**Comment by dims** [comment]
@mkumatag @lubinsz can you please try?

File: build/build-image/cross/Dockerfile
Comment (dims) [inline build/build-image/cross/Dockerfile]: @liggitt @lavalamp is a PROTOBUF_VERSION bump risky? problem is that aarch64 is not available in 3.0.2

Comment (liggitt) [reply build/build-image/cross/Dockerfile]: as far as I know, the risk would be in the changes in the files that got generated... it looks like there was no diff with this, is that correct?

Comment (dims) [reply build/build-image/cross/Dockerfile]: correct! none of the CI jobs yelled about any diff(s)

File: hack/lib/golang.sh
Comment (mkumatag) [inline hack/lib/golang.sh]: can we add ppc64le section as well?

**Review by mkumatag — CHANGES_REQUESTED** [review]
Please add changes required for ppc64le platform as well.

File: build/build-image/cross/Dockerfile
Comment (mkumatag) [inline build/build-image/cross/Dockerfile]: Please extend the support for `ppc64le` platform as well, here is the file: https://github.com/protocolbuffers/protobuf/releases/download/v3.7.0/protoc-3.7.0-linux-ppcle_64.zip

**Comment by dims** [comment]
@mkumatag can you please send me a diff/commit that i can patch in? (i am never sure of the strings from `uname -m` and the GOOS/GOARCH)

File: build/build-image/cross/Dockerfile
Comment (lavalamp) [reply build/build-image/cross/Dockerfile]: Agree w/ Jordan

Comment (ixdy) [inline build/build-image/cross/Dockerfile]: changes to the cross image really want a change to `build/build-image/cross/VERSION` too.

Comment (dims) [reply build/build-image/cross/Dockerfile]: ack doing it now

File: build/build-image/cross/Makefile
Comment (ixdy) [inline build/build-image/cross/Makefile]: also, if the image is arch-specific, it probably wants an arch-specific tag?

File: build/build-image/cross/Dockerfile
Comment (dims) [reply build/build-image/cross/Dockerfile]: so may be we should wait for @mkumatag too

/hold

**Comment by ixdy** [comment]
I think there's also an environment variable you need to set to prevent the build system from pulling the kube-cross image.

File: build/build-image/cross/Makefile
Comment (dims) [reply build/build-image/cross/Makefile]: and it's probably best to add the manifest as well!

Comment (mkumatag) [reply build/build-image/cross/Makefile]: I don't know how relevant to call this image as `kube-cross` for non-amd64 architecture because we aren't installing `crossbuild-essential-<platform>` for the non-amd64 platform so that way I really whether we will be able to build any binaries for different archs other than host arch.

File: build/build-image/cross/Dockerfile
Comment (mkumatag) [reply build/build-image/cross/Dockerfile]: and I see arch-specific etcd download as well in this file towards the end.

**Review by mkumatag — CHANGES_REQUESTED** [review]

**Comment by BenTheElder** [comment]
I think this LGTM, any more comments @ixdy ?

File: build/build-image/cross/VERSION
Comment (ixdy) [inline build/build-image/cross/VERSION]: the version is actually based on the underlying golang version, so this should be `v1.12.0-1`.

File: build/build-image/cross/Makefile
Comment (ixdy) [inline build/build-image/cross/Makefile]: I think we generally prefer using `docker manifest` now (e.g. the various `debian-*` images in `build/`)? though I guess this also works.

File: build/build-image/cross/VERSION
Comment (dims) [reply build/build-image/cross/VERSION]: Ack. switching it back

File: build/build-image/cross/Makefile
Comment (dims) [reply build/build-image/cross/Makefile]: Ack. let me switch this too

File: build/build-image/cross/Dockerfile
Comment (ixdy) [inline build/build-image/cross/Dockerfile]: where is this used?

File: build/build-image/cross/Dockerfile
Comment (ixdy) [inline build/build-image/cross/Dockerfile]: do we even need to cross-build from arm64 and ppc64le to other architectures?

File: build/build-image/cross/Dockerfile
Comment (ixdy) [inline build/build-image/cross/Dockerfile]: does this uname trick work if you're running this under qemu?

(also, don't we need to install qemu if building these images for arm or ppc on amd64?)

**Review by ixdy — COMMENTED** [review]
A few more comments. This basically looks fine, but I'm still not sure why we're doing dockerized builds on non-amd64 platforms.

File: build/build-image/cross/Dockerfile
Comment (dims) [reply build/build-image/cross/Dockerfile]: Dropped

File: build/build-image/cross/Dockerfile
Comment (dims) [reply build/build-image/cross/Dockerfile]: Didn't seem to be a problem, tried on MacOS, Ubuntu-x86, Ubuntu-arm64. Please let me know if you see a problem

Comment (mkumatag) [reply build/build-image/cross/Dockerfile]: I don't see copying the qemu binary anywhere in the flow! I really doubt whether we will be able to cross-build the docker image for other archs on linux/amd64 platform.

Cross build may happen on MacOS without issues because macos will support by default(more information - https://docs.docker.com/docker-for-mac/multi-arch/)

**Comment by soggiest** [comment]
@dims @liggitt @mkumatag This PR needs approval, can someone approve before code thaw?

**Comment by dims** [comment]
/milestone clear

let's not rush this now...it's too late for the release

**Comment by dims** [comment]
@mkumatag can you please lgtm?

File: build/build-image/cross/Makefile
Comment (mkumatag) [inline build/build-image/cross/Makefile]: Need to remove echo from this command to push actual manifest

**Review by mkumatag — CHANGES_REQUESTED** [review]
I ran the `make all` and pushed the manifest to my personal registry and what I found is below:

```
# uname -a
Linux localhost 4.4.0-142-generic #168-Ubuntu SMP Wed Jan 16 20:59:29 UTC 2019 ppc64le ppc64le ppc64le GNU/Linux


# docker run -ti mkumatag/kube-cross:v1.12.1-2 go version
standard_init_linux.go:190: exec user process caused "exec format error"
```

This is happening because we aren't building the kube-cross for other architectures properly, for this to work we need to take following things:

1. Add baseimage properly for every architecture properly instead of just `golang:1.12.1`
2. Add qemu-static to dockerfile dor respective architecture for cross building

**Comment by dims** [comment]
@mkumatag can you please kick the tires once more?

File: build/build-image/cross/Dockerfile
Comment (mkumatag) [inline build/build-image/cross/Dockerfile]: need to change amd64 to the specific arch folder.

**Review by mkumatag — CHANGES_REQUESTED** [review]
small correction needed, rest all seems okay to me.

File: build/build-image/cross/Dockerfile
Comment (dims) [reply build/build-image/cross/Dockerfile]: Ack doing that now

**Comment by dims** [comment]
@ixdy this looks ready now.

/test pull-kubernetes-verify

**Comment by mkumatag** [comment]
@dims I try building the container and see following error, can you please check what is happening:

```
Step 16/16 : RUN export ETCD_VERSION=v3.2.24;   mkdir -p /usr/local/src/etcd   && cd /usr/local/src/etcd   && curl -fsSL https://github.com/coreos/etcd/releases/download/${ETCD_VERSION}/etcd-${ETCD_VERSION}-linux-${ARCH}.tar.gz | tar -xz   && ln -s ../src/etcd/etcd-${ETCD_VERSION}-linux-${ARCH}/etcd /usr/local/bin/
 ---> Running in b61f09e75b1b
curl: (22) The requested URL returned error: 404 Not Found

gzip: stdin: unexpected end of file
tar: Child returned status 1
tar: Error is not recoverable: exiting now
The command '/bin/sh -c export ETCD_VERSION=v3.2.24;   mkdir -p /usr/local/src/etcd   && cd /usr/local/src/etcd   && curl -fsSL https://github.com/coreos/etcd/releases/download/${ETCD_VERSION}/etcd-${ETCD_VERSION}-linux-${ARCH}.tar.gz | tar -xz   && ln -s ../src/etcd/etcd-${ETCD_VERSION}-linux-${ARCH}/etcd /usr/local/bin/' returned a non-zero code: 2
Makefile:67: recipe for target '.container-amd64' failed
make[1]: *** [.container-amd64] Error 2
make[1]: Leaving directory '/root/k8s_ws/src/k8s.io/kubernetes/build/build-image/cross'
Makefile:56: recipe for target 'sub-container-amd64' failed
make: *** [sub-container-amd64] Error 2
root@jupiter-vm1278:~/k8s_ws/src/k8s.io/kubernetes/build/build-image/cross#
```

**Comment by dims** [comment]
@mkumatag i had to add another `ARG ARCH`, can you please try now?

**Comment by mkumatag** [comment]
/lgtm

```shell
root@icpscale220:~# docker run -ti mkumatag/kube-cross:v1.12.1-2
root@53b37f68b092:/go# which etcd
/usr/local/bin/etcd
root@53b37f68b092:/go# etcd --version
etcd Version: 3.2.24
Git SHA: 420a45226
Go Version: go1.8.7
Go OS/Arch: linux/ppc64le
root@53b37f68b092:/go# uname -a
Linux 53b37f68b092 4.4.0-145-generic #171-Ubuntu SMP Tue Mar 26 12:43:49 UTC 2019 ppc64le GNU/Linux
root@53b37f68b092:/go#
```

**Comment by cblecker** [comment]
/cc
/hold

Going to review this. I can already see a few things I'd like fixed up, but need a bit more time to go in depth.

File: build/build-image/cross/Makefile
Comment (cblecker) [inline build/build-image/cross/Makefile]: I don't like this. Building cross now requires root?

**Review by cblecker — CHANGES_REQUESTED** [review]

File: build/build-image/cross/Makefile
Comment (dims) [reply build/build-image/cross/Makefile]: we use this for the same thing for all the images ... :(

Comment (BenTheElder) [reply build/build-image/cross/Makefile]: cross compiling has always required this IIRC, just root in a container usually.

**Comment by mkumatag** [comment]
Looking forward to see this merged for kind project.

**Comment by cblecker** [comment]
/lgtm cancel

Other than the need for sudo (which I'm still thinking if there's a way around it), this mostly LGTM.

We need to bump the version in `build/build-image/cross/version` though, and then get the image built before we can merge.

Can we handle this after KubeCon? I'd like to make sure we have all the right eyes on this before we pull the trigger.

**Comment by dims** [comment]
@cblecker ack, builds are going to fail when i update `build/build-image/cross/VERSION` :)

yes we can handle this after kubecon EU.

**Comment by cblecker** [comment]
@dims Yes, but without that change, you aren't even testing this code path properly, so the "build" results are false.

**Comment by dims** [comment]
@cblecker yep, agree.

File: build/build-image/cross/Dockerfile
Comment (cblecker) [reply build/build-image/cross/Dockerfile]: I'm confused by this too.. we just don't need these packages if we aren't on a x86_64 machine?

File: build/build-image/cross/Dockerfile
Comment (cblecker) [inline build/build-image/cross/Dockerfile]:
```suggestion
ARG BASE_IMAGE=golang:1.12.5
```

File: build/build-image/cross/Makefile
Comment (cblecker) [reply build/build-image/cross/Makefile]: This breaks the build under Darwin:
```
# When cross-building, only the placeholder "CROSS_BUILD_" should be removed
# Register /usr/bin/qemu-ARCH-static as the handler for non-x86 binaries in the kernel
sudo ../../../third_party/multiarch/qemu-user-static/register/register.sh --reset
No binfmt support in the kernel.
  Try: '/sbin/modprobe binfmt_misc' from the host
make[1]: *** [.container-arm64] Error 1
make: *** [sub-container-arm64] Error 2
```

**Comment by cblecker** [comment]
Left a couple comments. Wasn't able to build this locally.

File: build/build-image/cross/Dockerfile
Comment (javier-b-perez) [inline build/build-image/cross/Dockerfile]: why QEMUARCH is not an ARG?

Comment (dims) [reply build/build-image/cross/Dockerfile]: @javier-b-perez please see the Makefile, we replace `QEMUARCH` using `sed`

**Comment by dims** [comment]
@cblecker - should be fixed now. am now running `make container` in `~/go/src/k8s.io/kubernetes/build/build-image/cross` to make sure.

**Comment by dims** [comment]
@cblecker - `make all-container` will not work on MacOS. (the code is the same as build/debian-base/Makefile)

File: build/build-image/cross/Dockerfile
Comment (BenTheElder) [reply build/build-image/cross/Dockerfile]: is there a reason we still do that? when I dug into this last time it seemed like it was just legacy to do this vs ARG

**Comment by cblecker** [comment]
This is still not a great build experience:
- the default make rule is for the manifest tool?
- `make container` works, but not `make all-container` on darwin? Why?
- Why do we need root on the build machine, as opposed to root in a linux container?
- there are dangling `.container-amd64` files.. should these targets be PHONY?

I hear that this is the same format as the hyperkube images, but all that means to me is that the hyperkube process is broken too.

**Comment by dims** [comment]
@mkumatag can you please respond to @cblecker 's concerns?

**Comment by cblecker** [comment]
Q: is this needed for 1.15? I'm happy to give review time to it if that's the intention, but it doesn't feel polished right now.

**Comment by spiffxp** [comment]
We're now basically at code freeze, this is going to need an exception if we're going to push on this

**Comment by dims** [comment]
Ack let's just move this out to 1.16

/milestone v1.16

**Comment by BenTheElder** [comment]
/retest
We're out of code freeze now :-)

**Comment by spiffxp** [comment]
need rebase

**Comment by dims** [comment]
/hold

( sorry i assumed that i had added the hold but it was @cblecker )

**Comment by cblecker** [comment]
Questions from my last review round: https://github.com/kubernetes/kubernetes/pull/75058#issuecomment-497067093

**Comment by dims** [comment]
@mkumatag since i am using the snippets of code that you used elsewhere in k/k, can you please help answer some of the questions from @cblecker ? (see previous comment from Christoph)

**Comment by spiffxp** [comment]
@cblecker FWIW this is my experience of trying to use `make all-container` on darwin without this PR, I'm not sure this PR makes anything more broken per se: https://github.com/kubernetes/kubernetes/pull/78396#issuecomment-503632093

**Comment by dims** [comment]
/close

we can reopen when @mkumatag or anyone else that needs this can work on it.

**Comment by chendave** [comment]
> This is still not a great build experience:
> * Why do we need root on the build machine, as opposed to root in a linux container?

Great question, I am recently working on the similar issue, and was redirected to this PR.
currently, the cross build need the root access to register interpreter for non-x86 binaries in the kernel, we need this cause some image building depends on this to fulfill some operations on build machine.

Take `echoserver` for example, the dockerfile looks like below,
```
FROM BASEIMAGE

CROSS_BUILD_COPY qemu-QEMUARCH-static /usr/bin/

RUN apk add --no-cache openssl nginx-mod-http-lua nginx-mod-http-lua-upstream
RUN mkdir -p /run/nginx

ADD nginx.conf /etc/nginx/nginx.conf
...
```
In the case of cross-build, take building ARM64 on AMD64 for example,
The command above "RUN" will need ARM64 interpreter (qemu-aarch64-static) to make the command able to run on AMD64 platform, without the code below,
```bash
$(SUDO) ../../../third_party/multiarch/qemu-user-static/register/register.sh --reset
```
You will see error message like this during the build process,
```
Step 3/10 : RUN apk add --no-cache openssl nginx-mod-http-lua nginx-mod-http-lua-upstream
 ---> Running in 86a69c7f0270
standard_init_linux.go:211: exec user process caused "exec format error"
The command '/bin/sh -c apk add --no-cache openssl nginx-mod-http-lua nginx-mod-http-lua-upstream' returned a non-zero code: 1
```
There is a chance to play with "docker in docker" or "docker out of docker" like what Moby does to avoid touching the ENV. on host,  I will give a shoot.

cc @dims  @mkumatag @cblecker @BenTheElder  @spiffxp

**Comment by chendave** [comment]
> > This is still not a great build experience:
> > 
> > * Why do we need root on the build machine, as opposed to root in a linux container?
> 
> Great question, I am recently working on the similar issue, and was redirected to this PR.
> currently, the cross build need the root access to register interpreter for non-x86 binaries in the kernel, we need this cause some image building depends on this to fulfill some operations on build machine.
> 
> Take `echoserver` for example, the dockerfile looks like below,
> 
> ```
> FROM BASEIMAGE
> 
> CROSS_BUILD_COPY qemu-QEMUARCH-static /usr/bin/
> 
> RUN apk add --no-cache openssl nginx-mod-http-lua nginx-mod-http-lua-upstream
> RUN mkdir -p /run/nginx
> 
> ADD nginx.conf /etc/nginx/nginx.conf
> ...
> ```
> 
> In the case of cross-build, take building ARM64 on AMD64 for example,
> The command above "RUN" will need ARM64 interpreter (qemu-aarch64-static) to make the command able to run on AMD64 platform, without the code below,
> 
> ```shell
> $(SUDO) ../../../third_party/multiarch/qemu-user-static/register/register.sh --reset
> ```
> 
> You will see error message like this during the build process,
> 
> ```
> Step 3/10 : RUN apk add --no-cache openssl nginx-mod-http-lua nginx-mod-http-lua-upstream
>  ---> Running in 86a69c7f0270
> standard_init_linux.go:211: exec user process caused "exec format error"
> The command '/bin/sh -c apk add --no-cache openssl nginx-mod-http-lua nginx-mod-http-lua-upstream' returned a non-zero code: 1
> ```
> 
> There is a chance to play with "docker in docker" or "docker out of docker" like what Moby does to avoid touching the ENV. on host, I will give a shoot.
> 
> cc @dims @mkumatag @cblecker @BenTheElder @spiffxp

we cannot do this in DinD or DooD cause this is a kernel feature, run root in container and do the `reset` is actually done in the host machine instead.

**Comment by BenTheElder** [comment]
binfmt_misc is a well known problem for us. I'm hoping we'll be moving these builds to GCB soon, which runs "containerized" but on a disposable per-build VM.

https://github.com/kubernetes/test-infra/issues/14091

**Pull request closed without merging**
